icefoganalytics/internal-data-portal
# Data Owner: Ability to Manage Dataset Fields

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/35

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/7

# Context

## Dataset Fields Manage Page - /dataset/:slug/fields/manage

### Front-end routes

Various tabs on the same level as this component.

### Back-end routes

* /api/dataset-fields - GET, list data set fields
* /api/dataset-fields - POST, create new data set fields
* /api/dataset-fields/:id - PATCH, update data set fields
* /api/dataset-fields/:id - DELETE, delete data set fields

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetFieldsManagePage.vue`
    * support some kind of prefill operation if Dataset#subscription_url exists, likely very high complexity
    * `FieldsEditTable.vue`
        * Standard edit table, with top level create action, and per-record update/delete actions

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/8779ef74-07ee-4a49-872e-b522a5a48524)

## DatasetField Model

`dataset_fields` table

### Fields

* `id`
* `dataset_id`
* `name`
* `display_name` - might be generated and non-editable
* `data_type` - an enum of things like `integer`, and `text`, not sure how many options are available.
* `description`
* `note`
* `created_at`
* `updated_at`
* `deleted_at`

### Relationships

* Belongs to a Dataset

# Implementation

Build a basic table with a top level create action and per-row update/delete actions.
Use modals for everything linked to query params, that trigger show/hide.

## Concerns

- In the future, it might be better to replace the modals with pages for each action as modals aren't very accessible, and provide a more limited experience.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3f5f5c4e-9798-45aa-98e4-7f19bb95932d)

> save button is cut of in screenshot, but it _is_ there
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3d5161b5-9161-4ba9-a1a9-303e615caf43)

> save button is cut of in screenshot, but it _is_ there
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/68d98078-42d9-4e44-8c41-3d75ba4a8b80)

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/aeb03d46-f1ab-4aaf-a191-e4ba90375204)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/07529118-7b2f-4e4a-80c8-d6a7b7c52fff)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Create a dataset from the dashboard, or go to the http://localhost:8080/datasets page and click on a dataset to edit it.
5. Note that there is now a Fields tab. When in "read" mode the tabs direct to the "read" views, when in "manage" mode, the tabs redirect to the "manage" pages.
6. Check that the breadcrumbs redirect to the expected locations depending on the tab you have selected.
7. Click the fields tab, then click the "edit' button to go to the fields manage table.
8. Add some fields.
9. Edit some fields.
10. Delete some fields.
11. Add more than 10 fields, and check that table pagination works correctly.
12. Attempt to add a field with the same name as a previous field, but with a different field display name. This will fail informatively.
13. Repeat this with a duplicated display name, and a different field name. It will still fail.
